### PR TITLE
SOF-2199 TransitionActions now sets the correct input source for VideoClips

### PR DESCRIPTION
**Pull request by LindvedKrvang**
VideoMixerTimelineObjectFactory can now create TimelineObjects with MediaPlayerSessions.

We use a placeholder as the input source for VideoClips since they need to be in our A/B roll and the correct input source won't be known untill execution (`OnTimelineGenerate` is where the actual input source is found). We use `mediaPlayerSession` to keep track of which Piece/TimelineObjects belongst to the same MediaPlayer.

In order for TransitionEffectActions to be find the correct input source they need to have a `mediaPlayerSession`. They now find the `mediaPlayerSession` of the Program Piece and adds it to its own TimelineObjects.

### CODE REVIEW

File: src/blueprints/tv2/action-factories/tv2-transition-effect-action-factory.ts [hunk 1]
```diff
@@ -350,54 +350,57 @@ export class Tv2TransitionEffectActionFactory extends ActionFactory {
       return action
     }
 
+    const metadata: Tv2PieceMetadata = piece.metadata as Tv2PieceMetadata
+    const mediaPlayerSession: string | undefined = metadata.mediaPlayerSessions && metadata.mediaPlayerSessions.length > 0 ? metadata.mediaPlayerSessions[0] : undefined
```
Comment (KvelaGorrrrnio) [inline src/blueprints/tv2/action-factories/tv2-transition-effect-action-factory.ts]: You can use optional chaining to simplify the expression:
```typescript
const mediaPlayerSession: string | undefined = metadata.mediaPlayerSessions?.[0]

**Review by KvelaGorrrrnio — APPROVED** [review]

File: src/blueprints/tv2/timeline-object-factories/tv2-atem-video-mixer-timeline-object-factory.ts [hunk 2]
```diff
@@ -289,25 +294,26 @@ export class Tv2AtemVideoMixerTimelineObjectFactory implements Tv2VideoMixerTime
...
-  private createTransitionEffectTimelineObject(layer: Tv2AtemLayer, meContent: AtemMeTimelineObject['content']['me']): AtemMeTimelineObject {
+  private createTransitionEffectTimelineObject(layer: Tv2AtemLayer, meContent: AtemMeTimelineObject['content']['me'], metadata?: Tv2TimelineObjectMetadata): AtemMeTimelineObject {
     return {
       id: `${layer}_${meContent.transition}`,
       enable: {
         start: 0
       },
       layer,
       priority: 10,
+      metaData: metadata,
```
Comment (MadsThy) [inline src/blueprints/tv2/timeline-object-factories/tv2-atem-video-mixer-timeline-object-factory.ts:316]:
```suggestion
      metadata,
```
Is it metaData or metadata? 😄 If we align the name, we won't need to assign it here like this, but could just do as the suggested here

File: src/blueprints/tv2/timeline-object-factories/tv2-tri-caster-video-mixer-timeline-object-factory.ts [hunk 3]
```diff
@@ -313,46 +323,61 @@ export class Tv2TriCasterVideoMixerTimelineObjectFactory implements Tv2VideoMixe
...
+  private createTransitionEffectTimelineObject(layer: Tv2TriCasterLayer, transitionEffectProperties: TriCasterTransitionEffectProperties, metadata?: Tv2TimelineObjectMetadata): TriCasterMixEffectTimelineObject {
     return {
-      id: `${layer}_${transitionEffect}`,
+      id: `${layer}_${transitionEffectProperties.transitionEffect}`,
       enable: {
         start: 0
       },
       layer,
       priority: 10,
+      metaData: metadata,
```
Comment (MadsThy) [inline src/blueprints/tv2/timeline-object-factories/tv2-tri-caster-video-mixer-timeline-object-factory.ts:345]: same as other comment about metadata :)

**Review by MadsThy — APPROVED** [review]

File: src/blueprints/tv2/timeline-object-factories/tv2-atem-video-mixer-timeline-object-factory.ts (hunk 2 above)
Comment (LindvedKrvang) [reply src/blueprints/tv2/timeline-object-factories/tv2-atem-video-mixer-timeline-object-factory.ts:316]: `metaData` is unfortunately how it is spelled in TSR. So in order for TSR to find the correct attribute, the attribute has to be named `metaData` on the `TimelineObject`.

File: src/blueprints/tv2/timeline-object-factories/tv2-tri-caster-video-mixer-timeline-object-factory.ts (hunk 3 above)
Comment (LindvedKrvang) [reply src/blueprints/tv2/timeline-object-factories/tv2-tri-caster-video-mixer-timeline-object-factory.ts:345]: Same reason as the other comment :)

File: src/blueprints/tv2/action-factories/tv2-transition-effect-action-factory.ts (hunk 1 above)
Comment (LindvedKrvang) [reply src/blueprints/tv2/action-factories/tv2-transition-effect-action-factory.ts]: Updated

**Review by RasmusAlbrektsen — APPROVED** [review]

**Review by KvelaGorrrrnio — APPROVED** [review]

**Pull request merged**